include/exadg/solvers_and_preconditioners/preconditioners/elementwise_preconditioners.h

```diff
@@ -192,6 +207,13 @@ class InverseMassPreconditioner
     integrator->reinit(cell);
   }
 
+  void
+  update() final
+  {
+    // no updates needed as long as the MatrixFree/Integrator object is up-to-date (which is not the
+    // responsibility of the present class).
+  }
```

Comment:
@kronbichler A few lines above in the constructor, there is the code

 ```C++
integrator = std::make_shared<Integrator>(matrix_free, dof_index, quad_index);
inverse = std::make_shared<CellwiseInverseMass>(*integrator);
```

I wonder whether these data structures are still correct if the mesh has been moved (with corresponding updates of matrix_free) in the background, or whether the constructor of `dealii::CellwiseInverseMass` would have to be called again? So it might be that the comment introduced for now is actually not correct.

Note that this is the `Elementwise::InverseMassPreconditioner`. For the "normal" or global `InverseMassOperator`, we create the `dealii::CellwiseInverseMass` inside the `cell_loop()`. We probably need to do this in a different way here in the `Elementwise` case, because we would otherwise create the object in every iteration of the elementwise Krylov solver (which is called for every cell). However, we could do some update inside the "global" `update()` function introduced in this PR, which is called once for all the cells (while the function `setup(cell)` then performs the setup related to a cell before the elementwise Krylov solver is called for every cell).

Reply:
@kronbichler what do you think on this topic? Should I wait merging this PR?

Reply:
Regarding correctness, the present code should be fine because `Integrator` (aka `FEEvaluation`) and `CellwiseInverseMatrix` do not hold a state regarding the geometry stored inside `MatrixFree`/`MappingInfo`, apart from pointers into these data structures. As long as only `MappingInfo::update_mapping()` (rather than `initialize()`) is called, these pointers should remain valid.

Regarding the general code design, it would be desirable to set up the temporary data such as `FEEvaluation` inside the loop over cells, where the call-back is made into the user code, and ideally have no `FEEvaluation` alive at all outside a matrix-free loop. This is because allowing objects to release their memory after use makes better use of data caches and, probably more importantly, avoids traversing long lists of data for use in `FEEvaluation` where the data is acquired (look at https://github.com/dealii/dealii/blob/4f8abe489f7ec609b88cfdf0d865cf1c9ee43b9c/include/deal.II/matrix_free/fe_evaluation.h#L2861 that calls https://github.com/dealii/dealii/blob/4f8abe489f7ec609b88cfdf0d865cf1c9ee43b9c/include/deal.II/matrix_free/matrix_free.h#L3039-L3052). Ideally, that list's length should be equal to the maximum of integrators present in a single cell or face work context. In that sense, every use case like the one via `shared_ptr` that persists across many calls is a use case we should strive to avoid. However, this is a long-standing design issue not related to the present change, and it would only be marginally be resolved by finding a new design here. However, we need to figure out a solution in a general re-design in the matrix-free loops that also includes how the deal.II functions are designed and how the call-backs look like. (I keep repeating this argument, so it should be well-known.)